### ISSUE TITLE
TTL Optimizer Feature

### ISSUE THREAD

**Issue by cbuijs** [folbricht/routedns#242]
For the `ttl-modifier`, having an "TTL optimizer" option to to modify TTL's to improve caching and make dns resolution more efficient, or at least have a couple of another ways to influence it.

So I would suggest to add an `optimize-ttl` (or better name like `modifier`) to `ttl-modifier`, that would take the following values (it will look at all the records in an anwer):

- `default` - The Default, also when `optmize-ttl` is ommitted (this is how `ttl-modifier` works today with `ttl-min` and `ttl-max` options).
- `lowest` - Take lowest TTL value
- `highest` - Take highest TTL value
- `average` - Take average of all TTL values
- `first` - Take TTL of first record
- `last` - Take TTL of last record

Optional in addition to above, as it is kind of weird/hazy one:
- `median` - Take TTL of median record (record in the middle)

Needs some more "thoughts":
- Have a `percentile` (separate) option that provides an option to take a percentage of the TTL values to work with. Ninety (90) percentage is common in some cases.

Consideration:
- The `ttl-min` and `ttl-max` still need to be respected, but... apply before taking TTL's or on the resulting/calculated TTL. I lean to the latter.

Basically all records in the answer will adopt the TTL "take" or "calculated" based on above options.

**Comment by folbricht** [comment]
I believe it's best to leave `ttl-min` and `ttl-max` to form a lower/upper bound on the value, with all the other functions being run before. That way it's always possible to keep values in check.
With regards to `percentile`, I'm not yet clear on what that calculation should look like. If 90%+ of records have the same value use that?

**Comment by folbricht** [comment]
There's a draft implementation on the `issue-242` branch. The value for `ttl-select` can be "lowest", "highest", "average", "first", "last". They should all be implemented, but I only tested on of them so far.

```
[groups.cloudflare-updated-ttl]
type = "ttl-modifier"
resolvers = ["cloudflare-dot"]
ttl-select = "average"
ttl-max = 86400
```

**Comment by cbuijs** [comment]
> I believe it's best to leave `ttl-min` and `ttl-max` to form a lower/upper bound on the value.

Yeah, that was my thinking as well, maybe not clear in my original message.

**Comment by cbuijs** [comment]
> There's a draft implementation on the `issue-242` branch. The value for `ttl-select` can be "lowest", "highest", "average", "first", "last". They should all be implemented, but I only tested on of them so far.

Awesome, testing today, will update later today with findings.

**Comment by cbuijs** [comment]
> With regards to `percentile`, I'm not yet clear on what that calculation should look like. If 90%+ of records have the same value use that?

No, it is simpler, but can be completly discarded. What I meant is that in most monitoring/timing systems, instead of an average, a 90% rule is used to be closer to reality and lost time processing to be more accurate. It is kind of mockup and representation of "loss" and show or use something closer to reality. Maybe over-engineered for this thing here ;-).

**Comment by cbuijs** [comment]
Some notes:

- Some of the comments/text in `ttl-modifier.go` still contain "average" on the non-average functions.
- When **no** `ttl-select` is provided, RouteDNS exits with message `ttl-select value: "`
- Some additional logging needed, we might want to add information "if" the TTL is adjusted, and to "what" value (and maybe also the ttl-select value).

**Comment by cbuijs** [comment]
Another suggestion:

`ttl-select = "random"`

It would select a random number between `ttl-min` and `ttl-max`.

**Comment by folbricht** [comment]
Fixed those two issues and added a "random" selector (not tested). The logging changes are going to be a bit more difficult.

**Comment by cbuijs** [comment]
Works as expected, please Merge.
